File: web-app/app/controllers/DemoController.scala
package controllers

import controllers.util.{ActivityOps, WebMetrics}
import model.apexcharts
import org.joda.time.LocalDate
import play.api.libs.json.Json
import play.api.mvc.{AbstractController, Action, AnyContent, ControllerComponents}
import velocorner.api.Units
import velocorner.api.chart.DailyPoint
import velocorner.api.wordcloud.WordCloud
import velocorner.model.DailyProgress
import velocorner.util.{DemoActivityUtils, JsonIo}

import javax.inject.Inject
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

class DemoController @Inject() (val connectivity: ConnectivitySettings, components: ControllerComponents)
    extends AbstractController(components)
    with ActivityOps
    with WebMetrics {

  // demo ytd data
  // route mapped to /api/demo/statistics/ytd/:action/:activity
  def ytdStatistics(action: String, activity: String): Action[AnyContent] = Action.async {
    timedRequest[AnyContent](s"demo ytd data for $action") { implicit request =>
      val activities = DemoActivityUtils.generate()
      val now = LocalDate.now()
      val series = toSumYtdSeries(activities, now, action, Units.Metric)
      Future(Ok(Json.obj("status" -> "OK", "series" -> Json.toJson(series))))
    }
  }

  // demo yearly progress data
  // route mapped to /api/demo/statistics/yearly/:action/:activity
  def yearlyStatistics(action: String, activity: String): Action[AnyContent] = Action.async {
    val activities = DemoActivityUtils.generate()
    val series = toYearlySeries(activities, action, Units.Metric)
    Future(Ok(Json.obj("status" -> "OK", "series" -> Json.toJson(series))))
  }

  // all daily activity list for the last 12 months, shown in the calendar heatmap
  // route mapped to /api/demo/statistics/daily/:action
  def dailyStatistics(action: String): Action[AnyContent] = Action.async { implicit request =>
    val activities = DemoActivityUtils.generate()
    val dailyProgress = DailyProgress.from(activities)
    val series = action.toLowerCase match {
      case "distance"  => dailyProgress.map(dp => DailyPoint(dp.day, dp.progress.to(Units.Metric).distance))
      case "elevation" => dailyProgress.map(dp => DailyPoint(dp.day, dp.progress.to(Units.Metric).elevation))
      case other       => sys.error(s"not supported action: $other")
    }
    Future(Ok(Json.toJson(series)))
  }

  // word cloud titles and descriptions with occurrences
  // route mapped to /api/demo/wordcloud
  def wordcloud(): Action[AnyContent] = Action { implicit request =>
    val wcs = DemoActivityUtils
      .generateTitles(200)
      .map(_.trim.toLowerCase)
      .groupBy(identity)
      .view
      .mapValues(_.size)
      .map { case (k, v) => WordCloud(k, v) }
      .toList
      .filter(_.weight > 1)
      .sortBy(_.weight)
      .reverse

    Ok(JsonIo.write(wcs))
  }

  // route mapped to /api/demo/statistics/histogram/:action/:activity
  def yearlyHistogram(action: String, activity: String): Action[AnyContent] = Action { implicit request =>
    val activities = DemoActivityUtils.generate()
    val series = action.toLowerCase match {
      case "distance"  => apexcharts.toDistanceHeatmap(activities, activity, Units.Metric)
      case "elevation" => apexcharts.toElevationHeatmap(activities, Units.Metric)
      case other       => sys.error(s"not supported action: $other")
    }
    Ok(Json.toJson(series))
  }
}

File: web-backend/src/main/kotlin/velocorner/backend/controller/DemoController.kt
package velocorner.backend.controller

import org.slf4j.LoggerFactory

import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import velocorner.backend.model.*
import velocorner.backend.util.ActivityOps.toSumYtdSeries
import velocorner.backend.util.ActivityOps.toYearlySeries
import velocorner.backend.util.DemoActivityUtil
import java.time.LocalDate

@RestController
@RequestMapping("/api/demo")
class DemoController {

    @GetMapping("/statistics/ytd/{action}/{activity}")
    fun ytdStatistics(
        @PathVariable action: String,
        @PathVariable activity: String
    ): ResponseEntity<Map<String, Any>> {
        val activities = DemoActivityUtil.generate()
        val now = LocalDate.now()
        val series = toSumYtdSeries(activities, now, action, Units.METRIC)

        return ResponseEntity.ok(
            mapOf(
                "status" to "OK",
                "series" to series
            )
        )
    }

    @GetMapping("/statistics/yearly/{action}/{activity}")
    fun yearlyStatistics(
        @PathVariable action: String,
        @PathVariable activity: String
    ): ResponseEntity<Map<String, Any>> {
        val activities = DemoActivityUtil.generate()
        val series = toYearlySeries(activities, action, Units.METRIC)

        return ResponseEntity.ok(
            mapOf(
                "status" to "OK",
                "series" to series
            )
        )
    }

    @GetMapping("/statistics/daily/{action}")
    fun dailyStatistics(
        @PathVariable action: String
    ): ResponseEntity<List<DailyPoint>> {
        val activities = DemoActivityUtil.generate()
        val dailyProgress = DailyProgress.from(activities)

        val series = when (action.lowercase()) {
            "distance" -> dailyProgress.map { dp ->
                DailyPoint(dp.day, dp.progress.to(Units.METRIC).distance)
            }

            "elevation" -> dailyProgress.map { dp ->
                DailyPoint(dp.day, dp.progress.to(Units.METRIC).elevation)
            }

            else -> throw IllegalArgumentException("Not supported action: $action")
        }

        return ResponseEntity.ok(series)
    }

    @GetMapping("/wordcloud")
    fun wordcloud(): ResponseEntity<List<WordCloud>> {
        val wordClouds = DemoActivityUtil
            .generateTitles(200)
            .map { it.trim().lowercase() }
            .groupingBy { it }
            .eachCount()
            .map { (word, count) -> WordCloud(word, count) }
            .filter { it.weight > 1 }
            .sortedByDescending { it.weight }

        return ResponseEntity.ok(wordClouds)
    }

    @GetMapping("/statistics/histogram/{action}/{activity}")
    fun yearlyHistogram(
        @PathVariable action: String,
        @PathVariable activity: String
    ): ResponseEntity<Any> {
        val activities = DemoActivityUtil.generate()
        val series = when (action.lowercase()) {
            "distance" -> Apexcharts.toDistanceHeatmap(activities, activity, Units.METRIC)
            "elevation" -> Apexcharts.toElevationHeatmap(activities, Units.METRIC)
            else -> throw IllegalArgumentException("Not supported action: $action")
        }

        return ResponseEntity.ok(series)
    }

    companion object {
        private val logger = LoggerFactory.getLogger(DemoController::class.java)
    }
}
